pkg/proxy/config/api.go

```diff
@@ -17,59 +17,140 @@ limitations under the License.
+		servicesChan <- ServiceUpdate{Op: ADD, Service: service}
 	}
-	if store == nil {
-		store = cache.NewStore(cache.MetaNamespaceKeyFunc)
+}
+
+func sendUpdateService(servicesChan chan<- ServiceUpdate) func(oldObj, newObj interface{}) {
+	return func(_, newObj interface{}) {
+		service, ok := newObj.(*api.Service)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("cannot convert to *api.Pod: %v", newObj))
+			return
+		}
+		servicesChan <- ServiceUpdate{Op: ADD, Service: service}
 	}
-	return &cache.UndeltaStore{
-		Store:    store,
-		PushFunc: fn,
+}
+
+func sendDeleteService(servicesChan chan<- ServiceUpdate) func(obj interface{}) {
+	return func(obj interface{}) {
+		var service *api.Service
+		switch t := obj.(type) {
+		case *api.Service:
+			service = t
+		case cache.DeletedFinalStateUnknown:
+			var ok bool
+			service, ok = t.Obj.(*api.Service)
+			if !ok {
+				utilruntime.HandleError(fmt.Errorf("cannot convert to *api.Pod: %v", t.Obj))
+				return
+			}
+		default:
+			utilruntime.HandleError(fmt.Errorf("cannot convert to *api.Pod: %v", t))
+			return
+		}
+		servicesChan <- ServiceUpdate{Op: REMOVE, Service: service}
 	}
 }
 
-// NewEndpointsStore creates an undelta store that expands updates to the store into
-// EndpointsUpdate events on the channel. If no store is passed, a default store will
-// be initialized. Allows reuse of a cache store across multiple components.
-func NewEndpointsStore(store cache.Store, ch chan<- EndpointsUpdate) cache.Store {
-	fn := func(objs []interface{}) {
-		var endpoints []api.Endpoints
-		for _, o := range objs {
-			endpoints = append(endpoints, *(o.(*api.Endpoints)))
+func sendAddEndpoints(endpointsChan chan<- EndpointsUpdate) func(obj interface{}) {
+	return func(obj interface{}) {
+		endpoints, ok := obj.(*api.Endpoints)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("cannot convert to *api.Endpoints: %v", obj))
+			return
 		}
-		ch <- EndpointsUpdate{Op: SET, Endpoints: endpoints}
+		endpointsChan <- EndpointsUpdate{Op: ADD, Endpoints: endpoints}
 	}
-	if store == nil {
-		store = cache.NewStore(cache.MetaNamespaceKeyFunc)
+}
+
+func sendUpdateEndpoints(endpointsChan chan<- EndpointsUpdate) func(oldObj, newObj interface{}) {
+	return func(_, newObj interface{}) {
+		endpoints, ok := newObj.(*api.Endpoints)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("cannot convert to *api.Endpoints: %v", newObj))
+			return
+		}
+		endpointsChan <- EndpointsUpdate{Op: ADD, Endpoints: endpoints}
 	}
-	return &cache.UndeltaStore{
-		Store:    store,
-		PushFunc: fn,
+}
+
+func sendDeleteEndpoints(endpointsChan chan<- EndpointsUpdate) func(obj interface{}) {
+	return func(obj interface{}) {
+		var endpoints *api.Endpoints
+		switch t := obj.(type) {
+		case *api.Endpoints:
+			endpoints = t
+		case cache.DeletedFinalStateUnknown:
+			var ok bool
+			endpoints, ok = t.Obj.(*api.Endpoints)
+			if !ok {
+				utilruntime.HandleError(fmt.Errorf("cannot convert to *api.Endpoints: %v", t.Obj))
```

Comment:
I don't understand. I think the only it can return is the cache.DeletedFinalStateUnknown.

I'm 100% sure that if this is not true, then we have at least 10+ places in the code which are not handling this one correctly. And a bunch of other different controllers won't be working correctly.